src/Peachpie.Library/Hash.cs

```diff
@@ -2618,16 +2622,264 @@ public static PhpString hash_pbkdf2(string algo, byte[] password, byte[] salt, i
+        public const int PASSWORD_ARGON2ID = 3;
+
+        #endregion
+
+        #region password_hash, password_verify, password_needs_rehash
+
+        private static bool CheckCost(PhpValue value, int lowerBound, int upperBound, string warnParseFail, string warnBoundFail, out int checkedValue)
+        {
+            if (value.IsInteger())
+            {
+                checkedValue = value.ToInt();
+                if (checkedValue >= lowerBound && checkedValue <= upperBound)
+                    return true;
+                else
+                {
+                    PhpException.Throw(PhpError.Warning, warnBoundFail);
+                    return false;
+                }
+            }
+            else
+            {
+                checkedValue = -1;
+                PhpException.Throw(PhpError.Warning, warnParseFail);
+                return false;
+            }
+        }
+
+        private static PhpValue HashPasswordArgon2(string password, PhpArray opt, bool argon2i_id)
+        {
+            // Default setting for argon2
+            int memory_cost = memory_costDefault;
+            int time_cost = costDefault;
+            int threads = threadsDefault;
+
+            if (opt != null)
+            {
+                // Argument memory cost for argon2
+                if (opt.ContainsKey("memory_cost") && (!CheckCost(opt.GetItemValue(new IntStringKey("memory_cost")), 4, int.MaxValue, Resources.LibResources.argon2_memory, Resources.LibResources.argon2_memory, out memory_cost)))
+                    return PhpValue.False;
+                // Argument time cost for argon2
+                if (opt.ContainsKey("time_cost") && !CheckCost(opt.GetItemValue(new IntStringKey("time_cost")), 1, int.MaxValue, Resources.LibResources.argon2_time, Resources.LibResources.argon2_time, out time_cost))
+                    return PhpValue.False;
+                // Argument threads for argon2
+                if (opt.ContainsKey("threads") && !CheckCost(opt.GetItemValue(new IntStringKey("threads")), 1, int.MaxValue, Resources.LibResources.argon2_threads, Resources.LibResources.argon2_threads, out threads))
+                    return PhpValue.False;
+            }
+            try
+            {
+                return HashArgon2(password, time_cost, memory_cost, threads, argon2i_id);
+            }
+            catch (Exception) { }
+
+            return PhpValue.False;
+        }
+
+        private static PhpValue HashPasswordBlowfish(string password,PhpArray opt)
+        {          
+            // Default setting for bcrypt
+            int cost = costDefault;
+            string salt = BCrypt.Net.BCrypt.GenerateSalt(cost);
+
+            if (opt != null)
+            {
+                // Argument cost for bcrypt
+                if (opt.ContainsKey("cost"))
+                {
+                    PhpValue costValue = opt.GetItemValue(new IntStringKey("cost"));
+                    if (costValue.IsInteger()) // Check right value of argument
```

Comment:
please replace with pattern
```c#
if (costValue.IsLong(out var costInt)) { ...
```
also consider what if `costValue` is `double` ? I suppose it would work in PHP, but we would throw a warning instead.